File: miss_net/lib/domain/entities/video.dart
import 'package:equatable/equatable.dart';

class Video extends Equatable {
  final String id;
  final String externalId;
  final String title;
  final String? coverUrl;
  final String sourceUrl;
  final DateTime createdAt;
  final String? duration;
  final String? releaseDate;
  final List<String>? actors;
  final List<String>? categories;
  final int? lastPositionMs;
  final int? totalDurationMs;

  const Video({
    required this.id,
    required this.externalId,
    required this.title,
    this.coverUrl,
    required this.sourceUrl,
    required this.createdAt,
    this.duration,
    this.releaseDate,
    this.actors,
    this.categories,
    this.lastPositionMs,
    this.totalDurationMs,
  });

  double get progress {
    if (lastPositionMs == null || totalDurationMs == null || totalDurationMs == 0) return 0.0;
    return (lastPositionMs! / totalDurationMs!).clamp(0.0, 1.0);
  }

  @override
  List<Object?> get props => [
        id,
        externalId,
        title,
        coverUrl,
        sourceUrl,
        createdAt,
        duration,
        releaseDate,
        actors,
        categories,
        lastPositionMs,
        totalDurationMs,
      ];
}

File: android-native/app/src/main/java/com/panyou/missnet/nativeapp/core/model/Video.kt
package com.panyou.missnet.nativeapp.core.model

import android.os.Parcelable
import androidx.compose.runtime.Immutable
import kotlinx.serialization.Serializable
import kotlinx.parcelize.Parcelize

@Parcelize
@Immutable
@Serializable
data class Video(
    val id: String,
    val externalId: String,
    val title: String,
    val coverUrl: String?,
    val sourceUrl: String,
    val createdAtEpochMs: Long,
    val duration: String? = null,
    val releaseDate: String? = null,
    val actors: List<String> = emptyList(),
    val categories: List<String> = emptyList(),
    val lastPositionMs: Long = 0L,
    val totalDurationMs: Long = 0L,
    val isOfflineReady: Boolean = false,
) : Parcelable {
    val progress: Float
        get() = if (totalDurationMs <= 0L) 0f else (lastPositionMs.toFloat() / totalDurationMs.toFloat()).coerceIn(0f, 1f)
}
